### DISCUSSION TITLE
Swapping Archaeologists

### DISCUSSION THREAD

**Discussion by adamgall** [sarcophagus-org/sarcophagus-v2-specs#6] [Ideas]
# What's the problem

In the first version of Sarcophagus, once an archaeologist has accepted a sarcophagus job, they're cryptographically tied to that sarcophagus until it's over.

No matter what happens (in the real world), the Sarcophagus system expects the archaeologist to be available whenever the sarcophagus' dead man's switch is flipped, else the sarcophagus social contract is broken, recipients won't receive their payload, and penalties are doled out.

# Let's fix it

A great solution to this problem would be the ability for an archaeologist to "opt out" of sarcophagus job that they're previously accepted, by transferring the rights and responsibilities of that sarcophagus to another archaeologist.

**Comment by dacebt** [comment]
Front-end: Currently the front-end only recognizes 'embalmers' and 'recipients'. We could update the WebApp to recognize Archaeologists accounts and provide an interface (similar flow to tomb for 'embalmers'). Allowing an Archaeologist to expand Sarcophagus and provide with a few actions.

Behind the scenes: transferring the right and responsibilities sounds a-lot like an ERC721. I believe we started this discussion before but This could provide a way to transfer a sarcophagus easily.

BrainDump:
- Would need to prevent transferring to non-archaeologist via contract
- Since there is no way to check if particular archaeologist is online or able to / should take on archaeologist ('should' being a archaeologist that has been marked for bad behavior. we would have to 'control' transfers in our market place or a whitelisted marketplace? (but this would still need to check for registered arches, as well as preventing any on chain transfers to happen (can we even do that? (or want to)).

**Reply by adamgall** [reply]
> Would need to prevent transferring to non-archaeologist via contract

Great question. My gut says no, so long as the system doesn't require any archaeologist-specific parameters to be known after a sarcophagus is created. For example, digging fees, which are set by the arch. When a sarcophagus is created, are digging fees for that sarcophagus static from that point forward? Or will they change if the underlying arch makes changes to their own rates.

**Reply by sethhrbek** [reply]
@adamgall -- could you clarify your thought process here?

My reading of @Da-Colon's point is that we need to enforce that when transferring the 1155, that the receiving address is a registered archaeologist (and not some rando non-arch addy).

**Reply by adamgall** [reply]
For sure, that was how I interpreted his message as well.

My response is: do we? What are the ramifications if the token is sent to a random (non archaeologist registered) address?

**Reply by dacebt** [reply]
If an archaeologist transfers ownership of a sarcophagus to a non-registered address, in other words a person who isn't running a archaeologist service, no service would be there to run to unwrapped that sarcophagus.

**Reply by adamgall** [reply]
1. Just because an address isn't "registered" as an archaeologist on-chain doesn't necessarily mean it's not running compatible archaeologist software that could complete an unwrapping event.

2. Should that (assumption) be a concern of the smart contracts?

**Reply by sethhrbek** [reply]
That would open the door to allow an entity that is *not* a registered archaeologist to do an unwrapping. Any entity that is not a registered archaeologist I would assign lower probability that unwrapping would occur.  Obv someone could register as an archaeologist *without* actually setting up and running the arch service, but I would put a higher probability that a registered arch address is running the arch service than a rando non-arch address.

Other ramifications: the reputation system --- we would need to track reputation of these outside entities if we need any kind of "minimum reputation threshold" when doing a transfer. This could get confusing for embalmers if they see a set of Archs they can choose when creating a Sarcophagus, then these other entities that they can't choose but have reputation.

What are the *benefits* of allowing non-registered addresses to own these? Besides freedom to blindly trade the token? V1 had a way for Archs to get paid at a different address than the one that was registered with (in case we still need that feature in v2).

**Reply by sethhrbek** [reply]
> Just because an address isn't "registered" as an archaeologist on-chain doesn't necessarily mean it's not running compatible archaeologist software that could complete an unwrapping event.

So is the fundamental difference here that "registered" archaeologists can be the "initial" archaeologists on a Sarcophagus, but then *anyone* can have the Sarcophagus transferred to them?

If the contract cares about the "initial" archaeologist, shouldn't it also care about the "transferred" one as well? "Caring" here meaning the arch must be registered on-chain.

**Reply by adamgall** [reply]
That's a really really good point.

So, yes, in my head I was thinking what you just said: initial set of archaeologists are registered on chain, but if rights&responsibilities are transferred, then maybe that new arch doesn't need to have been registered.

But you bring up a great point. That introduces complexity, the contract is now keeping track of different "kinds" of archaeologists (when bringing reputation into the mix).

So let me continue to make things more difficult:

In a well-connected p2p topology (where archaeologists are actually connected to each other, go with me on this for a moment), do archaeologists need to "register" themselves into the system?

Maybe simply spinning up an arch server could be enough.

**Reply by adamgall** [reply]
Then as long as an embalmer gets connected to one of those archaeologists, it can gossip it's way through the rest of the archaeologists, grabbing information about all of the along the way.

**Reply by sethhrbek** [reply]
I follow you, that makes sense. The contract could still track reputation metrics (i.e. tracking addresses that interact with certain methods) if need be, but it may not need to have a concept of "registered" archaeologists since the arch network can be determined off-chain.

We would need to dig into this a little more to see if this is feasible. In v1, a "registered" archaeologist has to put up bond if they want to be added to a sarcophagus -- only archs with sufficient free bond can be assigned to a sarcophagus. Would this bonding mechanic have to change in this new scenario? I'm struggling to decouple the concepts of being registered and having bond put up in the system.

**Reply by adamgall** [reply]
Bonding would definitely need to be thought through again. Maybe the simple action of "placing a bond" is enough to make yourself be considered as "registered". Rather, maybe to be a "valid archaeologist" you need to be both connected into the archaeologist network, and also have opened a bond in the system.

It very well may be that this whole discussion is a non starter, and we continue with the concept of "registered archaeologists", and include the contract constraint that "rights&responsibilities" NFTs can only be transferred to other registered archaeologists.

**Comment by adamgall** [comment]
Thinking more about the "archaeologist rights & responsibilities as ERC721/1155 tokens"...

Conceptually, I think it's a great idea. It allows other tooling (wallets, etc) to _at the very least_ display "archaeologist rights & responsibilities" natively.

I'm not sure if we need to worry about creating the UI to enable this, yet. First order of business is to design and build this into the contract layer.

Perhaps each new sarcophagus is represented as a new collection within an ERC1155 contract. Upon sarcophagus creation, 5 tokens (specifically, `n` tokens) would be minted in a new collection that is tied to the newly created sarcophagus, and assign the 5 selected archaeologists as owners of those tokens. During the unwrapping workflow, the system would just need to confirm that the caller (archaeologist, or whomever) owns one of those 5 tokens.

**Reply by adamgall** [reply]
This would also simplify digging fee payments. Fees would simply be paid to the addresses that own the NFTs.

**Reply by sethhrbek** [reply]
The Arch is responsible for unwrapping a specific shard (which has been encrypted with their private key) -- how does this responsibility get transferred? Would that private key (which in theory should only be used for the singular purpose of decrypting a shard) be transferred to the new arch somehow?

**Reply by sethhrbek** [reply]
Or could the shard be unencrypted by the sending arch, then re-encrypted by the receiving arch? Haven't thought this through much, spitballing here

**Reply by adamgall** [reply]
> Or could the shard be unencrypted by the sending arch, then re-encrypted by the receiving arch? Haven't thought this through much, spitballing here

Yes, this is how I was expecting it could work.

Not ideal, since the sending arch still "knows" the data from the shard. They can't use it within the context of the system, but are there other nefarious things that could happen?

A more "ideal" thing would be if the private key shard from the sending arch were completely invalidated/nullified. How might that happen? Maybe this is where [MPC](https://github.com/sarcophagus-org/sarcophagus-v2/discussions/2#discussioncomment-2463950) could come into play... a totally new shard were generated, to replace the old one (and on the contract, we mark the old shard as being unusable. Not sure if this is cryptographically possible though...

**Comment by adamgall** [comment]
Questions / concerns / comments for the group, which all kind of affect each other:

- can/should a "recusal" happen without having a recipient archaeologist at the ready? put another way: should there ever be a situation in which "rights and responsibilities" are not "owned" by anyone, able to be picked up by anyone? my gut says "no, if an archaeologist wants to recuse, there needs to be another archaeologist willing to take on those rights and responsibilities in real time". making this an "asynchronous" process would add a LOT of technical overhead... i'm not sure how it would be done.

- from a "system design" point of view, i think i have a pretty good idea of how this feature will be implemented at the core "rights and responsibilities are atomically transferred from one arch to another". however, i haven't yet brought the idea of a "marketplace" into my design. for example, can/should there be the concept that one arch can "sell" their (perhaps lucrative) "rights and responsibilities" for ETH/SARCO/stables/etc? should the whole concept of a marketplace be included here? my gut says that would be really cool, but probably not necessary to make the core mechanism work. we can likely build a "marketplace" contract and interface later, which allows "payment" to be a part of that atomic transaction

**Reply by USCMig** [reply]
> can/should a "recusal" happen without having a recipient archaeologist at the ready?

It can if we implement the `m` of `n` scheme to be an m of `n+1` then a given sarcophagus could have a single archaeologist recuse itself with no harm to the unwrapping, but this isn't sustainable or fair for other archaeologists given they may need to recuse as well... So maybe this is a non-starter and we just agree that when an archaeologist leaves one needs to replace it.

Handling the recusal process when removing the free bond is called:
- An arch chooses to recuse and meaning they are going to be cashing out and will need to remove their free bond locked in the system to call it a day… in that contract if they pull the free bond and have sarcophagi associated with it, the transaction that they are pulling their free bond they can pay for the transference of the shard they hold to another qualified archaeologist.  That process of finding a new one will need to be flushed out, but the recusing party will pay a "slashing/opt-out" penalty in covering the fees to transfer the shard to the new archaeologist.

The only caveat to the above scenario is that an archaeologist wants to recuse a sarcophagus but still wants to run the service (which I think is contradictory, but possible).

**Reply by adamgall** [reply]
> It can if we implement the m of n scheme to be an m of n+1 then a given sarcophagus could have a single archaeologist recuse itself with no harm to the unwrapping, but this isn't sustainable or fair for other archaeologists given they may need to recuse as well... So maybe this is a non-starter and we just agree that when an archaeologist leaves one needs to replace it.

To start, I think we're on the same page about this not being ideal. To further explain why this is not ideal: there's a big technical challenge with making this happen. If the embalmer creates one extra shard, but the future archaeologist for that shard is unknown, then that shard can't be encrypted (with an archaeologist's public key). So, it couldn't be stored publicly (on Arweave). Non-starter indeed.

Regarding removing free bond while recusing archaeologist duties, I'm initially approaching it as two independent actions.

1. Archaeologist recuses duties. All this does is "unlock" some locked bond, turning it back into free bond.
2. Archaeologist claws back their free bond. This action simply allows an archaeologist to pull any of their currently free bond out of the system.

Naively, if an archaeologist's goal is to exit the system _completely_, they need to perform those two steps _in order_.
